### ISSUE TITLE
rook-agent using default volume plugin path when a non-default has been set on the Kubelet

### ISSUE THREAD

**Issue by beatlejuse** [rook/rook#1162]
I update CoreOS from 1409.7.0 to 1520.8.0 and kubernters from 1.6.1 to 1.8.1 and now i have error when "rook-agent" conteiners running.
all (3) rook-agent conteiners have status RunContainerError.
And in conteiners logs i see this:
kubectl -n rook-system logs rook-agent-mklwd
failed to open log file "/var/log/pods/7a5eb042-be51-11e7-8ce2-0050569d04a3/rook-agent_1.log": open /var/log/pods/7a5eb042-be51-11e7-8ce2-0050569d04a3/rook-agent_1.log: no such file or directory
I can provide any data to solve this problem if necessary.

**Comment by jbw976** [comment]
@beatlejuse what does the following show (it will try to get the last termination message for the rook-agent pods)?
```
kubectl -n rook-system get pod -l app=rook-agent -o jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.status.containerStatuses[0].lastState.terminated.message}{"\n"}{end}'
```

**Comment by beatlejuse** [comment]
kubectl -n rook-system get pod -l app=rook-agent -o jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.status.containerStatuses[0].lastState.terminated.message}{"\n"}{end}'
rook-agent-c4wj8        mkdir /usr/libexec/kubernetes: read-only file system
rook-agent-fm64q        mkdir /usr/libexec/kubernetes: read-only file system
rook-agent-mklwd        mkdir /usr/libexec/kubernetes: read-only file system

**Comment by jbw976** [comment]
Thank you for gathering that information @beatlejuse.  On CoreOS, we've seen this before where the Kubernetes default for volume plug-ins isn't actually writeable.

I suspect this documentation will help you: https://rook.io/docs/rook/master/k8s-pre-reqs.html#flexvolume-configuration

Also this documentation I've been working on, but haven't released yet, could help you even more: https://github.com/jbw976/rook/blob/flexvolume-help/Documentation/common-problems.md#rook-agent-deployment

Let us know if you have any other questions or feedback on the new documentation I've been working on.  Thanks!

**Comment by beatlejuse** [comment]
Thank you!
i'm read the manual, but i'm already configure "volume-plugin-dir".
How can i ensure that this option is really works?

**Comment by kokhang** [comment]
@beatlejuse the value of `volume-plugin-dir` should be a directory in the host that is writable (as supposed to the default `/usr/libexec` which is not)
Also since kubelet in coreos runs in a container, it needs to be bind mounted. `/var/lib/kubelet` is normally bind mounted in coreos, so we recommend using this directory as `volume-plugin-dir`

**Comment by beatlejuse** [comment]
Yes, folder "/var/lib/kubelet/volumeplugins/" created and key "volume-plugin-dir" link to this folder.
I'm using this manual to up cluster: https://coreos.com/kubernetes/docs/1.6.1/deploy-master.html
I'm rebuild cluster and have no changes((
I think, new version of coreos (even hyperkube) can't recognize key "volume-plugin-dir"

**Comment by kokhang** [comment]
can you ensure that /var/lib/kubelet/volumeplugins is host mounted in the kubelet container? Can you try using `--volume-plugin-dir=/etc/kubernetes/volumeplugins` and see if that works?

**Comment by beatlejuse** [comment]
yes, i'm already tested folder "/etc/kubernetes/volumeplugins".
but in coreos kubelet started in rkt and i can't ensure that the folder is mounted because rkt don't have exec command.
i try run kubelet in docker, but this is too difficult for me.

**Comment by galexrt** [comment]
I'm running with Container Linux and made those changes to make the kubelet pickup the K8s volume plugin(s) for me.
I think you need to add the following lines to the `kubelet.service` you are using.
In the `ExecStartPre` block add the line:
```
ExecStartPre=/bin/mkdir -p /var/lib/kubelet/volumeplugins
```
and in the `ExecStart` command add:
```
--volume-plugin-dir=/var/lib/kubelet/volumeplugins
```

The `/var/lib/kubelet` directory gets mounted by the `kubelet-wrapper` command.

Could you try that and see if that works for you?

**Comment by beatlejuse** [comment]
yes, folders created and key "--volume-plugin-dir=/var/lib/kubelet/volumeplugins" include in kubelet-wrapper, but it dont take effect for me after update.
today i'm portable kubelet from rkt (kubelet-wrapper) to docker.
i run kubelet as hyperkube from quay.io/coreos/hyperkube:v1.8.1_coreos.0
adn now i cen exec to container and ensure as folders mounting
># ls -la  /var/lib/kubelet
>total 48
>drwxr-xr-x.  6 root root 4096 Nov  3 17:40 .
>drwxr-xr-x.  1 root root 4096 Nov  3 17:43 ..
>drwxr-xr-x.  2 root root 4096 Nov  3 17:40 pki
>drwxr-x---.  2 root root 4096 Nov  3 17:40 plugins
>drwxr-x---. 10 root root 4096 Nov  3 18:14 pods
>drwxr-xr-x.  2 root root 4096 Nov  3 17:39 volumeplugins
but i have the same error:
rook-agent-c4wj8 mkdir /usr/libexec/kubernetes: read-only file system

**Comment by jbw976** [comment]
@beatlejuse do the rook-agent pods need to get regenerated to pick up the new `--volume-plugin-dir` value from the Kubelet after it was restarted?  Described in recent docs here: https://rook.io/docs/rook/master/common-problems.html#rook-agent-deployment

**Comment by kokhang** [comment]
also all the kubelet has to be modified. If you are running multiple nodes, make sure you set this flag to all kubelet on all nodes and restart them. We know its a bit tedious currently on coreos but we are working into making that more pleasant :)

**Comment by beatlejuse** [comment]
No, its very easy for me! my cluster is boot and configure from PXE. Every time i erase all nodes and after 15 min i have a new cluster configurated by new configuration files)))

**Comment by beatlejuse** [comment]
yes, i set this key on all (3) nodes.
now i again tested both keys:  "/var/lib/kubelet/volumeplugins" and "/etc/kubernetes/volumeplugins"
and saw the same error.
this is logs from kubelet container:
`I1103 23:32:30.284421    1468 kuberuntime_manager.go:499] Container {Name:rook-agent Image:registry.xxx.ru/k8s/rook:latest Command:[] Args:[agent] WorkingDir: Ports:[] EnvFrom:[] Env:[{Name:POD_NAMESPACE Value: ValueFrom:&EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:metadata.namespace,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}} {Name:NODE_NAME Value: ValueFrom:&EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:spec.nodeName,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}}] Resources:{Limits:map[] Requests:map[]} VolumeMounts:[{Name:flexvolume ReadOnly:false MountPath:/flexmnt SubPath: MountPropagation:<nil>} {Name:dev ReadOnly:false MountPath:/dev SubPath: MountPropagation:<nil>} {Name:sys ReadOnly:false MountPath:/sys SubPath: MountPropagation:<nil>} {Name:libmodules ReadOnly:false MountPath:/lib/modules SubPath: MountPropagation:<nil>} {Name:modprobe ReadOnly:false MountPath:/sbin/modprobe SubPath: MountPropagation:<nil>} {Name:modinfo ReadOnly:false MountPath:/sbin/modinfo SubPath: MountPropagation:<nil>} {Name:rook-operator-token-m772r ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath: MountPropagation:<nil>}] LivenessProbe:nil ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:Always SecurityContext:&SecurityContext{Capabilities:nil,Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:nil,} Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we should restart it.
I1103 23:32:30.284656    1468 kuberuntime_manager.go:738] checking backoff for container "rook-agent" in pod "rook-agent-lj9c9_rook-system(e2732084-c0d4-11e7-a1f5-0050569d04a3)"
I1103 23:32:30.284940    1468 kuberuntime_manager.go:748] Back-off 5m0s restarting failed container=rook-agent pod=rook-agent-lj9c9_rook-system(e2732084-c0d4-11e7-a1f5-0050569d04a3)
E1103 23:32:30.285028    1468 pod_workers.go:182] Error syncing pod e2732084-c0d4-11e7-a1f5-0050569d04a3 ("rook-agent-lj9c9_rook-system(e2732084-c0d4-11e7-a1f5-0050569d04a3)"), skipping: failed to "StartContainer" for "rook-agent" with CrashLoopBackOff: "Back-off 5m0s restarting failed container=rook-agent pod=rook-agent-lj9c9_rook-system(e2732084-c0d4-11e7-a1f5-0050569d04a3)"`
What else can i check?

**Comment by beatlejuse** [comment]
@galexrt are you shure that you are using kubernetes 1.8.1 and coreoc 1520.8.0 ?
its not work for me.

**Comment by galexrt** [comment]
@beatlejuse I'm using the current Tectonic `master` with K8s `v1.7.9` and Container Linux `1520.8.0`.

**Comment by beatlejuse** [comment]
@galexrt  it's amazing!
in first message i wrote that i get error after update to version 1.8.1
in older version i have not get this error ))
can you test this in kubernetes 1.8.1 ?

**Comment by kahkhang** [comment]
I believe I'm having the same error as well when testing on kubernetes 1.8.1. Here is the PR i am trying to do kahkhang/kube-linode#53

**Comment by galexrt** [comment]
@beatlejuse I just upgrade my cluster to K8s `1.8.2`. I had to add the env var `FLEXVOLUME_DIR_PATH: /var/lib/kubelet/volumeplugins` (with the volume plugins path) to the rook operator or else the operator started the agents with the default path `/usr/libexec/...`.
```
env:
[...]
- name: FLEXVOLUME_DIR_PATH
          value: "/var/lib/kubelet/volumeplugins"
[...]
```

**Comment by beatlejuse** [comment]
@galexrt yes, it;s working! Thank you!

**Comment by kahkhang** [comment]
Can confirm this is working too :) Maybe the extra environment variable should be documented somewhere or perhaps put into the manifest https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/rook-operator.yaml

**Comment by jbw976** [comment]
Something doesn't seem right here still.  If you told the Kubelet to use `/var/lib/kubelet/volumeplugins` then at runtime, when the operator queries its node's configz endpoint, then that same value should be returned.  You shouldn't have to tell the operator explicitly to use that path via its `FLEXVOLUME_DIR_PATH ` env var at all because the configz API should tell the operator the right value.  This is still very odd.

@kokhang do you have any ideas on why the configz value would still come back as the default even though it looks like the Kubelet has correctly been told to use a non-default path?

@galexrt do you happen to have your operator logs from before you added the `FLEXVOLUME_DIR_PATH` env var?

**Comment by galexrt** [comment]
@jbw976 I don't have the exact logs, but the operator was telling that it defaulted to the operator default of `FLEXVOLUME_DIR_PATH` (`/usr/libexec/...`) instead of using my volume plugins flag value.

**Comment by tailtwo** [comment]
Same problem with CoreOS 1520.8.0, kubernetes 1.8.3 and rook 0.6.0. I'm using the kubelet flag like this : `--volume-plugin-dir=/var/lib/kubelet/volumeplugins` but it gets ignored as I can see in the rook-agent daemon set JSON :
```json
        "volumes": [
          {
            "name": "flexvolume",
            "hostPath": {
              "path": "/usr/libexec/kubernetes/kubelet-plugins/volume/exec/",
              "type": ""
            }
```

**Comment by jbw976** [comment]
@tailtwo can you share your rook-operator pod logs? I am curious what it's seeing from its perspective when it is asking kubernetes what path to use for plugins?

**Comment by kokhang** [comment]
@kahkhang @tailtwo Can you execute the following `kubectl get --raw=/api/v1/nodes/<nodename>/proxy/configz`? And tell me the value of `volumePluginDir`?

@jbw976 you are right. Rook should auto discover this value from the /configz api. I am not sure why this is happen in that CoreOS environment since this has worked on other CoreOS... One thing i can think of is that Rook doesnt have access or priviledge to the /configz api.

**Comment by kokhang** [comment]
@paha just ran into this when he deployed K8s in GCE.

**Comment by tailtwo** [comment]
@jbw976 

Here is my operator logs : 

```bash
2017-11-16 08:55:39.822515 I | rook: starting Rook v0.6.0 with arguments '/usr/local/bin/rook operator'
2017-11-16 08:55:39.822614 I | rook: flag values: --help=false, --log-level=INFO, --mon-healthcheck-interval=20s, --mon-out-timeout=5m0s
2017-11-16 08:55:39.916976 I | rook: starting operator
2017-11-16 08:55:39.936867 I | opkit: creating cluster CRD
2017-11-16 08:55:39.943335 I | opkit: CRD cluster already exists
2017-11-16 08:55:39.943358 I | opkit: creating pool CRD
2017-11-16 08:55:39.958290 I | opkit: CRD pool already exists
2017-11-16 08:55:39.958307 I | opkit: creating objectstore CRD
2017-11-16 08:55:39.970797 I | opkit: CRD objectstore already exists
2017-11-16 08:55:39.970813 I | opkit: creating filesystem CRD
2017-11-16 08:55:39.989278 I | opkit: CRD filesystem already exists
2017-11-16 08:55:39.989293 I | opkit: creating volumeattachment CRD
2017-11-16 08:55:40.016087 I | opkit: CRD volumeattachment already exists
2017-11-16 08:55:42.641293 I | op-k8sutil: cluster role rook-agent already exists. Updating if needed.
2017-11-16 08:55:42.739988 I | op-agent: getting flexvolume dir path from provided env var
2017-11-16 08:55:42.740005 I | op-agent: flexvolume dir path is not provided. Defaulting to: /usr/libexec/kubernetes/kubelet-plugins/volume/exec/
2017-11-16 08:55:42.740011 I | op-agent: using flexvolume dir path: /usr/libexec/kubernetes/kubelet-plugins/volume/exec/
2017-11-16 08:55:42.750938 I | op-agent: rook-agent daemonset started
2017-11-16 08:55:42.817905 I | operator: rook-provisioner started
2017-11-16 08:55:42.818289 I | opkit: start watching cluster resource in all namespaces at v1alpha1
```

Also, here is a sample of my kubelet log during the helm deployment.

```bash
Nov 16 08:55:36 node1 kubelet-wrapper[975]: I1116 08:55:36.789438     975 reconciler.go:212] operationExecutor.VerifyControllerAttachedVolume started for volume "rook-operator-token-mx6bs" (UniqueName: "kubernetes.io/secret/e9a4098f-caab-11e7-847f-3440b581e460-rook-operator-token-mx6bs") pod "rook-operator-6c8b87dcd7-z5kt6" (UID: "e9a4098f-caab-11e7-847f-3440b581e460")
Nov 16 08:55:42 node1 kubelet-wrapper[975]: E1116 08:55:42.785688     975 helpers.go:468] PercpuUsage had 0 cpus, but the actual number is 16; ignoring extra CPUs
Nov 16 08:55:42 node1 kubelet-wrapper[975]: I1116 08:55:42.903506     975 reconciler.go:212] operationExecutor.VerifyControllerAttachedVolume started for volume "flexvolume" (UniqueName: "kubernetes.io/host-path/ed4e3b4a-caab-11e7-847f-3440b581e460-flexvolume") pod "rook-agent-z84p7" (UID: "ed4e3b4a-caab-11e7-847f-3440b581e460")
Nov 16 08:55:42 node1 kubelet-wrapper[975]: I1116 08:55:42.903554     975 reconciler.go:212] operationExecutor.VerifyControllerAttachedVolume started for volume "libmodules" (UniqueName: "kubernetes.io/host-path/ed4e3b4a-caab-11e7-847f-3440b581e460-libmodules") pod "rook-agent-z84p7" (UID: "ed4e3b4a-caab-11e7-847f-3440b581e460")
Nov 16 08:55:42 node1 kubelet-wrapper[975]: I1116 08:55:42.903590     975 reconciler.go:212] operationExecutor.VerifyControllerAttachedVolume started for volume "modinfo" (UniqueName: "kubernetes.io/host-path/ed4e3b4a-caab-11e7-847f-3440b581e460-modinfo") pod "rook-agent-z84p7" (UID: "ed4e3b4a-caab-11e7-847f-3440b581e460")
Nov 16 08:55:42 node1 kubelet-wrapper[975]: I1116 08:55:42.903624     975 reconciler.go:212] operationExecutor.VerifyControllerAttachedVolume started for volume "dev" (UniqueName: "kubernetes.io/host-path/ed4e3b4a-caab-11e7-847f-3440b581e460-dev") pod "rook-agent-z84p7" (UID: "ed4e3b4a-caab-11e7-847f-3440b581e460")
Nov 16 08:55:42 node1 kubelet-wrapper[975]: I1116 08:55:42.903673     975 reconciler.go:212] operationExecutor.VerifyControllerAttachedVolume started for volume "sys" (UniqueName: "kubernetes.io/host-path/ed4e3b4a-caab-11e7-847f-3440b581e460-sys") pod "rook-agent-z84p7" (UID: "ed4e3b4a-caab-11e7-847f-3440b581e460")
Nov 16 08:55:42 node1 kubelet-wrapper[975]: I1116 08:55:42.903712     975 reconciler.go:212] operationExecutor.VerifyControllerAttachedVolume started for volume "rook-agent-token-7q7kv" (UniqueName: "kubernetes.io/secret/ed4e3b4a-caab-11e7-847f-3440b581e460-rook-agent-token-7q7kv") pod "rook-agent-z84p7" (UID: "ed4e3b4a-caab-11e7-847f-3440b581e460")
Nov 16 08:55:42 node1 kubelet-wrapper[975]: I1116 08:55:42.903747     975 reconciler.go:212] operationExecutor.VerifyControllerAttachedVolume started for volume "modprobe" (UniqueName: "kubernetes.io/host-path/ed4e3b4a-caab-11e7-847f-3440b581e460-modprobe") pod "rook-agent-z84p7" (UID: "ed4e3b4a-caab-11e7-847f-3440b581e460")
Nov 16 08:55:43 node1 kubelet-wrapper[975]: E1116 08:55:43.970121     975 remote_runtime.go:209] StartContainer "860517399d881c74cc79511323e9174fe4d1f9e1b410e8753d7b7ef8c0698e67" from runtime service failed: rpc error: code = Unknown desc = failed to start container "860517399d881c74cc79511323e9174fe4d1f9e1b410e8753d7b7ef8c0698e67": Error response from daemon: mkdir /usr/libexec/kubernetes: read-only file system
Nov 16 08:55:43 node1 kubelet-wrapper[975]: E1116 08:55:43.970201     975 kuberuntime_manager.go:715] container start failed: RunContainerError: failed to start container "860517399d881c74cc79511323e9174fe4d1f9e1b410e8753d7b7ef8c0698e67": Error response from daemon: mkdir /usr/libexec/kubernetes: read-only file system
Nov 16 08:55:43 node1 kubelet-wrapper[975]: E1116 08:55:43.970237     975 pod_workers.go:182] Error syncing pod ed4e3b4a-caab-11e7-847f-3440b581e460 ("rook-agent-z84p7_rook-system(ed4e3b4a-caab-11e7-847f-3440b581e460)"), skipping: failed to "StartContainer" for "rook-agent" with RunContainerError: "failed to start container \"860517399d881c74cc79511323e9174fe4d1f9e1b410e8753d7b7ef8c0698e67\": Error response from daemon: mkdir /usr/libexec/kubernetes: read-only file system"
Nov 16 08:55:45 node1 kubelet-wrapper[975]: W1116 08:55:45.137621     975 helpers.go:847] eviction manager: no observation found for eviction signal allocatableNodeFs.available
Nov 16 08:55:45 node1 kubelet-wrapper[975]: I1116 08:55:45.177518     975 kuberuntime_manager.go:500] Container {Name:rook-agent Image:rook/rook:v0.6.0 Command:[] Args:[agent] WorkingDir: Ports:[] EnvFrom:[] Env:[{Name:POD_NAMESPACE Value: ValueFrom:&EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:metadata.namespace,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}} {Name:NODE_NAME Value: ValueFrom:&EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:spec.nodeName,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}}] Resources:{Limits:map[] Requests:map[]} VolumeMounts:[{Name:flexvolume ReadOnly:false MountPath:/flexmnt SubPath: MountPropagation:<nil>} {Name:dev ReadOnly:false MountPath:/dev SubPath: MountPropagation:<nil>} {Name:sys ReadOnly:false MountPath:/sys SubPath: MountPropagation:<nil>} {Name:libmodules ReadOnly:false MountPath:/lib/modules SubPath: MountPropagation:<nil>} {Name:modprobe ReadOnly:false MountPath:/sbin/modprobe SubPath: MountPropagation:<nil>} {Name:modinfo ReadOnly:false MountPath:/sbin/modinfo SubPath: MountPropagation:<nil>} {Name:rook-agent-token-7q7kv ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath: MountPropagation:<nil>}] LivenessProbe:nil ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:IfNotPresent SecurityContext:&SecurityContext{Capabilities:nil,Privileged:*true,SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:nil,} Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we should restart it.
Nov 16 08:55:45 node1 kubelet-wrapper[975]: I1116 08:55:45.177608     975 kuberuntime_manager.go:739] checking backoff for container "rook-agent" in pod "rook-agent-z84p7_rook-system(ed4e3b4a-caab-11e7-847f-3440b581e460)"
Nov 16 08:55:45 node1 kubelet-wrapper[975]: E1116 08:55:45.425363     975 remote_runtime.go:209] StartContainer "a83b698b5471c43b282566d21a70fe313e02b4fc5a8f862e8fd38d7382669d6f" from runtime service failed: rpc error: code = Unknown desc = failed to start container "a83b698b5471c43b282566d21a70fe313e02b4fc5a8f862e8fd38d7382669d6f": Error response from daemon: mkdir /usr/libexec/kubernetes: read-only file system
Nov 16 08:55:45 node1 kubelet-wrapper[975]: E1116 08:55:45.425415     975 kuberuntime_manager.go:715] container start failed: RunContainerError: failed to start container "a83b698b5471c43b282566d21a70fe313e02b4fc5a8f862e8fd38d7382669d6f": Error response from daemon: mkdir /usr/libexec/kubernetes: read-only file system
Nov 16 08:55:45 node1 kubelet-wrapper[975]: E1116 08:55:45.425446     975 pod_workers.go:182] Error syncing pod ed4e3b4a-caab-11e7-847f-3440b581e460 ("rook-agent-z84p7_rook-system(ed4e3b4a-caab-11e7-847f-3440b581e460)"), skipping: failed to "StartContainer" for "rook-agent" with RunContainerError: "failed to start container \"a83b698b5471c43b282566d21a70fe313e02b4fc5a8f862e8fd38d7382669d6f\": Error response from daemon: mkdir /usr/libexec/kubernetes: read-only file system"
```

@kokhang This is the result on my 3 node cluster : 
Node 1 (slave) : `"volumePluginDir":"/var/lib/kubelet/volumeplugins"`
Node 2 (slave) : `"volumePluginDir":"/var/lib/kubelet/volumeplugins"`
Node 2 (slave) : `"volumePluginDir":"/var/lib/kubelet/volumeplugins"`

**Comment by bassam** [comment]
consider fixing in v0.6.1

**Comment by jbw976** [comment]
Root cause for this issue appears to be that with the dynamic kubelet config work in 1.8, the type returned from the configz API has been changed, so our marshaling is not working for 1.8.

Old type: https://github.com/kubernetes/kubernetes/blob/release-1.7/pkg/apis/componentconfig/types.go#L316
New type: https://github.com/kubernetes/kubernetes/blob/release-1.8/pkg/kubelet/apis/kubeletconfig/types.go#L187

Tricky part is that we aren't yet consuming any k8s packages from 1.8+, so we do not have this type readily accessible to build against.  That work is ongoing in #1153 

Here is a temporary hack, where we duplicate just the 1 field that we are looking to marshal in a 1.8+ environment: https://github.com/jbw976/rook/commit/d46a7fb3c7057c77b8d247a8c776baa96d282e1b#diff-81d2693aefad51616fc360533223e965R44

**Comment by travisn** [comment]
Great work @jbw976. Rather than , this change looks good for a couple reasons:
- Would work across kubernetes versions
- Simple to backport to 0.6.1 rather wait for the big work item #1153, which we don't want to backport

**Comment by bassam** [comment]
so this was a breaking change in 1.8? Also there were concerns about it only happening for 1.8.1+, do we have clarity on that now?

**Comment by jbw976** [comment]
Yes @bassam, this was a breaking change for 1.8.0 (all 1.8.x versions are affected), but it is somewhat up to debate because I don't think the configz API is formally documented.  We may want to think about a way to improve that going forward, but in the meantime I want to understand why we didn't hit this in our tests.  Perhaps in our test environment, the default under `/usr/libexec/...` just works and that is why we didn't see this.

**Comment by bassam** [comment]
How did we pass integration testing on 1.8? Do we have gaps in testing?

**Comment by Ulexus** [comment]
This is only an issue with the non-default location, which might explain the coverage gap.

**Comment by jbw976** [comment]
This operator log file from an integration test on 1.8 actually does show us hitting this condition: https://jenkins.rook.io/job/rook/job/rook/job/master/362/artifact/_output/tests/TestSmokeSuite_gce_1.8_rook-operator-544c5df9dc-65x9l_smoke-ns-system_1509561923.log

however, falling back to the default location actually works in that environment (we can install to it and the Kubelet will be looking for plugins there), so everything still works OK.  If I recall correctly, I think the images used for integration tests are based on Ubuntu?

I'd like some sort of integration verification to protect us from regressing this again, so I'm trying to think through how to do that for all versions of K8s.

**Comment by jbw976** [comment]
This has been fixed in both master and the release-0.6 branch.

**Comment by kokhang** [comment]
@jbw976 this issue has come back with k8s 1.9.x. The node/proxy query doesnt give you the flexvolume dir. I think some of this work was refactor when adding the dynamic kubelet config support. Ill reopen this issue.

**Comment by jbw976** [comment]
A fix for k8s 1.9 will be tracked in #1330 where it was first observed that we can't detect the flexvolume dir at runtime in 1.9.

**Comment by EnigmaCurry** [comment]
This is broken in rancher/k3os - the ENV var fix for FLEX_VOLUME_DIR_PATH in operator.yaml worked though - setting it to  `/var/lib/rancher/k3s/agent/kubelet/plugins` - https://github.com/rook/rook/issues/1162#issuecomment-343743163
